bem bench command

unlok opened the pull request.

Commit:
Build non resource demanding techs inprocess (close #250)

Build non resource demanding techs in the main process instead of the forked one. Return techs to build using fork() in BundleNode->getForkedTechs(), others will be built inprocess.

arikon committed:
Add ability to customize build rules more flexibly by providing Arch.createCustomNode() method

Commit:
Tests bem-bl submodule is updated to 0.2 (close #264)

- tests bem-bl updated
- tests fixed to inherit i18n.js from js-i to build files with include directives (close #265)

arikon committed:
Add match*() methods to `simple` level prototype, add tests (Close #282)

Commit:
Magic nodes doesn't link the nodes it creates with parent magic nodes
(reference #306)

scf2k committed:
Merge pull request #354 from bem/i-353

deps.js: don't swallow parsing errors (close #353)

Commit:
0.5.26

Commit:
0.5.26

Commit:
Fix bem make on node 0.10.x (close #357)

Commit:
Merge branch 'i-357'
Fix bem make on node 0.10.x (close #357)

Commit:
Add fork support to bem create command

Commit:
0.5.27

Commit:
Move borschik dependency (close #358)

Commit:
Add node 0.10 builds in travis

Commit:
0.5.28

Commit:
GitLibraryNode: don't update from upstream if working copy state satisfies to the configured one (close #335)

Commit:
0.5.29

Commit:
Merge branch 'i-335'
0.5.29

arikon committed:
Add changelog for 0.5.30

arikon committed:
0.5.30

arikon committed:
Merge branch 'release-v0.5.30'

Commit:
Techs V2 research

Commit:
ie.css tech should pass absolute path for its chunks

Commit:
level caches, tech build cache

Commit:
Fix for "Coud not call for method of undefined" when using nodes from
API

Commit:
- Use object model level cache

scf2k committed:
Merge pull request #378 from bem/bug-fn-undefined

Fix for "Coud not call for method of undefined" when using nodes from API

scf2k committed:
Merge pull request #379 from bem/bug-ie.css

ie.css tech should pass absolute path for its chunks

Commit:
- Add mkdrip wrapper to util.js

Commit:
Add additional techs and levels from abandoned introspect branch

Commit:
0.5.31

Commit:
Q.nbind -> Q.nfbind hotfix

Commit:
Merge branch 'release-v0.5.31'
0.5.31

Commit:
Creating parent directory for SymlinkLibraryNode if it doesn't exists (fix #342)

scf2k committed:
Merge pull request #384 from bem/issues-342

Create parent directory for SymlinkLibraryNode if one doesn't exist yet

Commit:
- getBuildFiles() now returns only the files which apply for the tech.
  Files are grouped by the build suffixes.
- getBuildFiles() fixed to filter the files in the levels order.

Commit:
- deps.js: use v2 api

Commit:
Init realisation of bench command

Commit:
Realisation:  extract each treeish

Commit:
Gets realpath and check exist file

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,69 @@
+
+
+// Namespace
+var BENCH = {
```
Для чего нужен неймспейс?

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,69 @@
+var cmd_exec  = require('child_process').exec;
```
По нашему code style один `var` и несколько переменных через запятую.

``` js
var a = '...',
    b = '...';
```

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,69 @@
+};
+
+
+BENCH.MODULES.TreeishExtractor = (function(){
```
Для чего здесь нужно замыкание?

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,69 @@
+    var treeish   = ['HEAD'].concat(args['treeish-list']);
+    //var benchmarks = opts.benchmarks;
+
+    treeish.forEach(function(trish) {
```
Здесь нужно собрать все промисы про чекаут из гита и вернуть из фунции через `Q.all([...])`

arikon replied:
``` js
return treeish.map(...)
```

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,69 @@
+    }
+
+    function cleanUp () {
+        cmd_exec('rm -rf tmp/*', function(err, strout, stderr) {
```
Здесь нужно вернуть promise

gurugray commented on lib/bench.js:
```diff
@@ -0,0 +1,69 @@
+            console.log('git checkout: '.green, path_to_treeish.underline.green);
+            
+            var level = require('./level').createLevel(path_to_treeish + '/benchmark.bundles/');
+            var path = level.getPath(level.getByObj({block:'b-link'}), 'bemjson.js');
```
тут `b-link` причём?

gurugray commented on lib/bench.js:
```diff
@@ -0,0 +1,69 @@
+module.exports = function(opts, args){
+
+    var treeish   = ['HEAD'].concat(args['treeish-list']);
+    //var benchmarks = opts.benchmarks;
```
мёртвый код

Commit:
do continuous promise of chain to COA

Commit:
Add wrapper on run cmd and do making

Commit:
Add proto benchmark testing

Commit:
Without rejection for normal debuging

Commit:
Refactoring all with worked flow

arikon committed:
bem create level command was failing on Node.js 0.10.x

Fix #372

arikon committed:
Merge branch 'i372'

arikon committed:
Changelog for 0.5.32

arikon committed:
0.5.32

arikon committed:
Merge branch 'release-0.5.32'

arikon committed:
Build for node 0.10 and 0.11

arikon committed:
Merge branch 'travis-node-0.10'

Commit:
Fix run of benchmarks for correct tests result

Commit:
Copy latest benchmarks to all dir

Commit:
Add -b param for specific benchmarks

Commit:
Add -w param for run without current state

Commit:
- Update to Q 0.9.4 and disable long stack capture

Commit:
- Delete items from the times array used in LOGGER.time()
- Add LOGGER.timeEndLevel() to allow specify logging level

Commit:
- Get rid of QQ

Commit:
- ie.css V2 use V2 validation

Commit:
- deps V2 use validation

Commit:
- Add combined one regexp for level matching

Commit:
- Both V1 and V2 techs can be used

Commit:
Using logger for output information

Commit:
- Small optimizations

Commit:
- Code cleanup

Commit:
Optimization make for specific benchmarks

Commit:
- Update q, borschik, xjst, ometajs

Commit:
Merge branch 'release-0.5.33'
0.5.33

Commit:
- Fix level and tech tests

Commit:
- deps V2 fixed to work in a forked process

Commit:
- Level: don't fail on level scan if directory does not exist

Commit:
Add output results in table

Commit:
Fix cloning etalon benchmarks

Commit:
Add cleaner temporary folder

Commit:
Add check of stable state

Commit:
Add rme calculating

Commit:
- Level: fix crashing logging about non existing level

Commit:
- Add a feature which allows to specify which levels to cache and which do not
- Add force-cache argument to the bem build command to avoid levels rescan when executed in a forked process

Commit:
- Fix getBuildSuffixesMap to return proper data for ie.css V2 tech

Commit:
- Refactor deps V2 and base tech V2 validation code to use promises chain

Commit:
- Store meta.js V2 techs files in the root/.bem/cache

Commit:
- deps.js V2: fix consequence dependant code

Commit:
- level: tech resolving is expanded to be able to find the tech in the bem/lib/techs/ folder and inner

Commit:
- Use getTech() instead of createTech() to benefit from the techs caching

Commit:
Add ONLY_BENCH mode and result sorting

Commit:
- Level cache exception can be a path containing the level(s)

Commit:
- Level scanning refactored to be faster
- getItemsByIntrospection() uses new level scanning when called with no custom opts

Commit:
Add -w  command. Fix result testing - add active waiting

Commit:
- V2 tech: files collecting changed from levels->dep to dep->levels

Commit:
- Move V2 techs to lib/techs/v2

Commit:
- Update tests to use V2 techs

Commit:
- Move bem-bl submodule pointer to the version with V2 techs

Commit:
- Fix tests after rebase

Commit:
- bem server: reset non-cacheable levels when there is no concurrent requests

Commit:
- V2 techs api is refactored a little

Commit:
- level test fixed to ensure items order in the getItemsByIntrospection() result

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+    BENCHMARK   = require('benchmark'),
+    VM          = require('vm');
+    UTIL        = require('./util'),
+    CHILD_PROC  = require('child_process'),
```
`CHILD_PROC` -> `CP`

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+    FS          = require('fs'),
+    BENCHMARK   = require('benchmark'),
+    VM          = require('vm');
+    UTIL        = require('./util'),
```
`UTIL` -> `U`

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+    CHILD_PROC  = require('child_process'),
+    LEVEL       = require('./level'),
+    TABLE      = require('cli-table'),
+    LOGGER      = require('./logger');
```
Можно сгруппировать модули — сначала глобальные, потом локальные

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+/**
+ * BemBenchmark - testing on speed  BEMHTML templates between revision
+ */
+function BemBenchmark(opts, args){
```
Я бы переименовал класс в `Benchmark`

arikon replied:
Используй модуль `INHERIT` для определение класса

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+ */
+function BemBenchmark(opts, args){
+
+    this.PATH_TMP        = 'bench/';
```
Свойства объекта называй как локальные переменные: `pathSelf`, `waiting` и т.д.

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+
+    rsync = CHILD_PROC.spawn('rsync', ['-av','--exclude=node_modules','--exclude=bench','--exclude=bem-bl','--exclude=.git','.',self.PATH_TMP + self.PATH_SELF]);
+
+    rsync.stdout.on('data', function(data){
```
Убери код, если он не используется

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+    }
+
+    // if -b flag not detected then all benchmarks
+    try {
```
Для чего здесь блок `try {} catch {}`, если ошибка кидается сразу же после обработки?

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+
+    var self = this,
+        level =  LEVEL.createLevel(treesh_path + level_path),
+        bemjson_files = [],
```
CODE STYLE: `bemjsonFiles`, `bemEntities`

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+        throw new Error(e);
+    }
+
+    if (bem_entities) {
```
Здесь проверка на truthy, выше (`self.BENCHMARKS`) на не `undefined`. Не единообразно.

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+                LOGGER.timeEnd('[' + target.green + '] has been assembled');
+                return [target].concat(links);
+            })
+            .fail(function(err) {
```
Если в обработчике `.fail()` не вернуть `Q.reject()`, то при ошибке цепочка промисов не прервётся. Это действительно то, что нужно в этом случае?

unlok replied:
Да. Так как один из проектов может просто не собраться, в таком случае лучше просто пропустить его и тестировать те проекты, которые соберутся.

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+
+        LOGGER.info('Benchmark...');
+        return Q.all(links_pack.map(function(pack){
+        var links  = pack.slice(1),
```
CODE STYLE: не хватает отступа для блока кода

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+ * 
+ * @return {} | none
+ */
+BemBenchmark.prototype.beckupRevision = function() {
```
`beckupRevision` -> `backupRevision`

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+
+            var def = Q.defer();
+            cmd = 'cp -R ' + self.PATH_TMP + (self.WITH_CURRENT ? self.LATEST_REVISION : self.PATH_SELF) +
+                '/' + self.PATH_BENCHMARKS + '* ' + self.PATH_TMP + target + '/' + self.PATH_BENCHMARKS;
```
Используй `PATH.join()` вместо конкатенации строк для построения путей

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+    LOGGER.time('All time');
+
+    if (!self.ONLY_BENCH) {
+        Q.when(self.cleaner())
```
Не хватает `return`

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+                    allBeckups = self.beckupRevision();
+                }
+
+                return Q.all(allBeckups)
```
Отсюда и ниже сложный для чтения кусок, слишком большая вложенность — ничем не лучше callback style.
Точно можно упростить.

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+                    .then(function() {
+                        return [target].concat(self.getBemjsonPath(self.PATH_TMP + target + '/' , self.PATH_BENCHMARKS));
+                    });
+            return def;
```
Почему не просто `return Q.resolve([target].concat(self.getBemjsonPath(self.PATH_TMP + target + '/' , self.PATH_BENCHMARKS)))`?

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+            self.activeWait(self.WAITING);
+
+            return self.bench(links, trg)
+                .then(function(res) {
```
Лишний `.then()`

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+            }));
+        })
+        .then(function(res) {
+             LOGGER.timeEnd('All time');
```
Лишний пробел

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+ */
+module.exports = function(opts, args) {
+    return new BemBenchmark(opts, args);
+};
```
Не хватает `\n` в последней строке файла

arikon commented on lib/commands/bench.js:
```diff
@@ -0,0 +1,47 @@
+                .end()
+            .opt()
+                .name('only-bench')
+                .short('ob').long('only-bench')
```
В `short()` можно передать только одну букву

arikon commented on lib/commands/bench.js:
```diff
@@ -0,0 +1,47 @@
+                .end()
+            .opt()
+                .name('delta-rme')
+                .short('dr').long('delta-rme')
```
В `short()` можно передать только одну букву

arikon commented on lib/commands/bench.js:
```diff
@@ -0,0 +1,47 @@
+                .end()
+            .opt()
+                .name('wait')
+                .short('wait').long('wait')
```
В `short()` можно передать только одну букву. Если не нужен `short`, можно его просто не использовать.

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+
+    var self = this,
+        def  = Q.defer(),
+        allBeckups,
```
`allBeckups` -> `allBackups`

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+        table.push(row);
+    });
+
+    console.log(table.toString());
```
Почему здесь не `return`, а вывод сразу в консоль?

arikon commented on lib/commands/bench.js:
```diff
@@ -0,0 +1,47 @@
+            .opt()
+                .name('delta-rme')
+                .short('dr').long('delta-rme')
+                .arr()
```
Для чего здесь `arr()`? Вроде не нужен.

arikon commented on lib/commands/bench.js:
```diff
@@ -0,0 +1,47 @@
+            .opt()
+                .name('wait')
+                .short('wait').long('wait')
+                .arr()
```
Для чего здесь `arr()`? Вроде не нужен.

arikon commented on lib/commands/bench.js:
```diff
@@ -0,0 +1,47 @@
+            .opt()
+                .name('without-current-state')
+                .short('w').long('wcs')
+                .arr()
```
Для чего здесь `arr()`? Вроде не нужен.

arikon commented on lib/commands/bench.js:
```diff
@@ -0,0 +1,47 @@
+            .opt()
+                .name('only-bench')
+                .short('ob').long('only-bench')
+                .arr()
```
Для чего здесь `arr()`? Вроде не нужен.

arikon commented on lib/commands/bench.js:
```diff
@@ -0,0 +1,47 @@
+            .title('Benchmarks BEMHTML').helpful()
+            .arg()
+                .name('treeish-list')
+                .title('List of treeish(revision) to compare')
```
Не хватает пробела перед `(revisions)`

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+/**
+ * Cleaning all repo from temporary folder 
+ * 
+ * @return {none}
```
`@return {Promise * Undefined}` (если ничего не возвращает)

arikon replied:
В других похожих местах аналогично

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+
+        LOGGER.time('[' + target.green + ' => ' +name_short.blue+ '] has been tested');
+
+        (function() {
```
Добавь комментарий про прогрев

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+        })();
+
+        suite.add(name, function() {
+            bemhtml.BEMHTML.apply(res);
```
Всё таки два теста или «активное» ожидание?

arikon commented on lib/bench.js:
```diff
@@ -0,0 +1,667 @@
+    });
+
+    return def.promise;
+};
```
CODE STYLE: Если есть пустая строка в начале функции (она нужна, если тело функции большое), нужно добавлять пустую строку в конце

Commit:
Add clear .md file to docs/

arikon commented on docs/bem-bench/bem-bench.ru.md:
```diff
@@ -0,0 +1,206 @@
+***
+##bem bench - тестирование скорости шаблонов
```
Не хватает пробела после `##`

arikon commented on docs/bem-bench/bem-bench.ru.md:
```diff
@@ -0,0 +1,206 @@
+
+Результатами тестирования является `HZ` - кол-во выполнений в единицу времени(в данном случае в секунду), а так же `RME` - относительная погрешность выраженная процентно от среднего арифметического, по этой величине можно судить, насколько истины результаты тестирования.
+
+####Этолонные benchmarks:
```
«Эталонные»

Двоеточие в заголовке не нужно

arikon commented on docs/bem-bench/bem-bench.ru.md:
````diff
@@ -0,0 +1,206 @@
+```
+*Запуск* **bem bench**:
+  
+```js
````
Это же не js блок кода. Можно не указывать форматтер, если не надо расцвечивать код.

arikon commented on docs/bem-bench/bem-bench.ru.md:
```diff
@@ -0,0 +1,206 @@
+
+*Результаты*:	
+
+	┌───┬───────────┬───────────────────────┬──────────────┬────────────────┬──────────┐
```
Стоит завернуть в ```, чтобы не ползло при рендеринге

arikon commented:
https://github.com/bem/bem-tools/blob/bench/docs/bem-bench/bem-bench.ru.md

Посмотри здесь на форматирование документа. Поправь всё, что выгладит криво.
С формулировками некоторых предложений стоит поработать — можно привлечь кого-то из документаторов, или вычитаем вместе ближе к вечеру.

Commit:
Code refactoring

Commit:
Add doc content and image

Commit:
Fix style

Commit:
Add project configuration

Commit:
v0.6.0

Commit:
Merge branch 'techs-v2'
v0.6.0

Commit:
Fix slyle and correct content

Commit:
Fix make

Commit:
Correct doc - change commands title and configuration way

Commit:
Fix error handling

Commit:
Fix code style and doc style

arikon committed:
Minor rewrite and fix formatting

Commit:
- add new scanning methods to levels/simple

arikon committed:
CODE REVIEW: Code style and jsdoc fixes, rename some vars

arikon committed:
Add dependencies from colors, cli-table and benchmark modules

arikon committed:
CODE REVIEW: Reject on `git archive` error

arikon committed:
CODE REVIEW: Reject on `bem make` error

arikon committed:
CODE REVIEW: Refactor and simplify `benchmarksCloner()` method

arikon committed:
CODE REVIEW: Reject on `git show` error in `getLatestTreeish()` method

Commit:
Refactoring 'cleaner'

Commit:
Refactoring 'cleaner'

Commit:
Rename ONLY BENCHMARKS mode

Commit:
Fix 'cleaner' - return promise

Commit:
Add function - 'defineBackups' and return cli-table to COA

arikon committed:
CODE REVIEW: Style and minor code fixes

arikon commented:
@unlok Саша, твои последний коммит ломает код. Разобраться и починить я не смог — потратил час, код очень запутан. В понедельник-вторник будем делать ревью вместе.

arikon commented:
Сейчас не работает с такой ошибкой.

```
arikon@arikon-osx:~/projects/bem/project-stub (bem-bench)$ bem bench -w HEAD HEAD~1 HEAD~2
11:13:53.334 - info: Cleaning a TMP folder
11:13:53.346 - info: TMP folder has been cleaned: 9ms
11:13:53.347 - info: Exclude current_state
11:13:53.369 - info: Create revision [HEAD~1]: 15ms
11:13:53.370 - info: Create revision [HEAD~2]: 12ms
11:13:53.372 - info: Create revision [HEAD]: 23ms
11:13:53.407 - info: Latest revision is - HEAD~1
11:13:53.427 - error: benchmarksCloner - Error: cp -R bench/HEAD~1/benchmark.bundles* bench/HEAD/benchmark.bundles failed: cp: bench/HEAD~1/benchmark.bundles*: No such file or directory

11:13:53.429 - error: benchmarksCloner - Error: cp -R bench/HEAD~1/benchmark.bundles* bench/HEAD~2/benchmark.bundles failed:
Error: cp -R bench/HEAD~1/benchmark.bundles* bench/HEAD/benchmark.bundles failed: cp: bench/HEAD~1/benchmark.bundles*: No such file or directory

    at ChildProcess.<anonymous> (/Users/arikon/projects/bem/bem-tools/lib/util.js:794:18)
    at ChildProcess.EventEmitter.emit (events.js:99:17)
    at Process._handle.onexit (child_process.js:686:10)
From previous event:
    at Object.exports.exec (/Users/arikon/projects/bem/bem-tools/lib/util.js:789:15)
    at INHERIT.benchmarksCloner (/Users/arikon/projects/bem/bem-tools/lib/bench.js:359:30)
```

В коде нигде не вызывается `this.backupSelf()`

Commit:
- write level cache (files.json) to level/.bem/cache dir

Commit:
level: add new scan methods to levels/simple so that it works with V2 techs
level: move level caches to lebel/.bem/cache

Commit:
Remove 'defineBackups' function and refactoring 'start'

unlok commented:
Сережа, это я ошибочно вынес функционал в функцию  defineBackups, логику которой не дописал. Сегодня пересмотрел, это оказалось лишним, вернул назад и сделал рефакторинг. Сейчас все работает.

Commit:
- Fix ieX.css V2 techs to write contents into correct output file

arikon committed:
benchmarkNoFounded -> benchmarksNotFound

arikon committed:
gitIterator -> exportGitRevision

arikon committed:
backupSelf -> exportWorkingCopy

arikon committed:
backupRevision -> exportGitRevisions

arikon committed:
cleaner() -> cleanTempDir()

arikon committed:
generateTable() -> getResultsTable()

arikon committed:
CODE REVIEW: Code style fixes

arikon committed:
Replace `CP.spawn()` and `CP.exec()` calls with `util.exec()` promised equivalent

arikon committed:
CODE REVIEW: Refactor of `getBemjsonPath()` method

- rename to `getBenchmarksPaths()`
- do not check for `.bemjson.js` file existence
- use PATH.join()
- do not store inexistent files in `this.benchmarksNotFound` property

Commit:
Merge branch 'cssv2-fix'

arikon committed:
Add error handling for `tar -x` command in `exportGitRevision()` method

arikon committed:
Remove `this.treeishListChecked` property

arikon committed:
Change `this.pathTmp` to `.bem/cache/bench`

Commit:
Move a benchmarking process to main flow 'start()'

Commit:
Move bem-bl submodule hash

arikon committed:
CODE REVIEW: Remove `this.latestRevision` property

arikon committed:
CODE REVIEW: benchmarksCloner() -> cloneBenchmarks()

arikon committed:
CODE REVIEW: Refactor `start()` and `makeTarget()` (former `maker()`) methods to simplify code

Commit:
- Fix path.resolve() call in level to not pass undefined
- Update to mocha 1.11 to make it work on node 0.10

Commit:
Rename methods - 'make()' to 'makeAll()', 'bench()' to 'runBenchmark()'

arikon committed:
Allow `bem bench` to be run without arguments, it will benchmark only working copy then

arikon committed:
Refactor of `bem bench` command line options and fix of docs

arikon committed:
Fix of exclude patterns for `rsync` on working copy export

Commit:
Merge branch 'node-10'
Fix node 0.10 failures

arikon committed:
Update `npm-shrinkwrap.json`

arikon committed:
Install latest `npm` on Travis build

arikon commented:
Close in favor of #393
